drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,495 @@
+	nonMachineConfiguredRules := []network.SecurityRule{}
+	for _, rule := range e.Rules {
+		if _, err := ConvertRuleToOpenPort(rule); err != nil {
+			name := to.String(rule.Name)
+			if name != r.SSHRule.Name && name != r.DockerRule.Name {
+				nonMachineConfiguredRules = append(nonMachineConfiguredRules, rule) // not a Rancher configured rule
+			}
+		}
+	}
+	if len(nonMachineConfiguredRules) > 0 {
+		for _, rule := range nonMachineConfiguredRules {
+			if err := checkIfDockerConflict(rule, r.DockerRule); err != nil {
+				return err
+			}
+		}
+		log.Warnf("Identified existing Azure security rules that are not configured through Rancher: %v",
+			securityRulesToLogField(nonMachineConfiguredRules))
+	}
+	return nil
+}
+
+// checkIfDockerConflict does a rudimentary check to see if the rule could potentially conflict with the ability
+// for the docker port to be accessible. If the rule does not pass all rudimentary checks, it returns an error.
+func checkIfDockerConflict(rule network.SecurityRule, dockerRule AllowInboundSecurityRule) error {
+	f := securityRuleToLogField(rule)
+	portInPortRange := func(port int, portRange string) (bool, error) {
+		if portRange == "*" {
+			return true, nil
+		}
+		if s := strings.Split(portRange, "-"); len(s) == 2 {
+			portMin, err := strconv.Atoi(s[0])
+			if err != nil {
+				return false, fmt.Errorf("Error parsing portMin for port range %s in rule: %s", portRange, f)
+			}
+			portMax, err := strconv.Atoi(s[1])
+			if err != nil {
+				return false, fmt.Errorf("Error parsing portMax for port range %s in rule: %s", portRange, f)
+			}
+			if portMin > portMax {
+				return false, fmt.Errorf("Unable to interpret port range %s with portMin > portMax in rule: %s", portRange, f)
+			}
+			return port >= portMin && port <= portMax, nil
+		} else if len(s) == 1 {
+			portVal, err := strconv.Atoi(s[0])
+			if err != nil {
+				return false, fmt.Errorf("Error parsing port %s in rule: %s", portVal, f)
+			}
+			return port == portVal, nil
+		} else {
+			return false, fmt.Errorf("Recieved invalid port range %s from rule: %s", portRange, f)
+		}
+	}
+	priority := to.Int32(rule.Priority)
+	if priority > dockerRule.Priority {
+		return nil
+	}
+	purpose := strings.ToLower(fmt.Sprintf("%s %s %s", rule.Access, rule.Direction, rule.Protocol))
+	if purpose == "deny inbound tcp" || purpose == "deny inbound *" {
+		// check if docker port is covered by rule
+		dockerPort, err := strconv.Atoi(dockerRule.DstPort)
+		if err != nil {
+			return fmt.Errorf("Cannot process non-integer dockerPort %s", dockerRule.DstPort)
+		}
+		deniedPortRanges := getDestinationPorts(rule)
+		for _, pr := range deniedPortRanges {
+			if portWithinRange, err := portInPortRange(dockerPort, pr); err != nil {
+				return err
+			} else if portWithinRange {
+				return fmt.Errorf("Detected an existing security rule that is set to %s requests to the Docker port %s: %s",
+					purpose, dockerRule.DstPort, f)
+			}
+		}
+	}
```

Comment:
We currently fail on create if a user has any existing security rule that has all the following:
- a higher priority than the DockerAllowAny rule (i.e. between 101 <= priority <= 299)
- Denies inbound tcp/* requests
- Contains the dockerPort in the DestinationPortRanges

Examples:
![blocking_docker_port_with_rules_0](https://user-images.githubusercontent.com/13127443/77803230-e2e00380-7039-11ea-8c27-d3541a7dea47.png)
![blocking_docker_port_with_rules_1](https://user-images.githubusercontent.com/13127443/77803236-e4113080-7039-11ea-8fe7-3be99aab8b21.png)
![blocking_docker_port_with_rules_2](https://user-images.githubusercontent.com/13127443/77803240-e5425d80-7039-11ea-9390-373b0326653d.png)
![blocking_docker_port_with_rules_3](https://user-images.githubusercontent.com/13127443/77803244-e5daf400-7039-11ea-9dc1-52786206bc31.png)

Such rules would would conflict with the way we set the Docker rule settings, shown below:

```
# Output from the r.DockerRule.Rule(), where r is the RancherMachineSecurityRuleConfiguration

network.SecurityRule{
 	Name: to.StringPtr("DockerAllowAny"),
 	SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
 		Description:              to.StringPtr("Allow docker engine access (TLS-protected)"),
 		SourceAddressPrefix:      to.StringPtr("*"),
 		DestinationAddressPrefix: to.StringPtr("*"),
 		SourcePortRange:          to.StringPtr("*"),
 		DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", dockerPort)),
 		Access:                   network.SecurityRuleAccessAllow,
 		Direction:                network.SecurityRuleDirectionInbound,
 		Protocol:                 network.SecurityRuleProtocolTCP,
 		Priority:                 to.Int32Ptr(300),
 	},
 }
```

However, based on the [`SecurityRulePropertiesFormat` Azure docs](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-11-01/network?tab=doc#SecurityRulePropertiesFormat), the user could theoretically have a rule that denies inbound tcp requests to the dockerPort (and would fail on create like the above situations), but only to specific:
- SourcePortRange(s)
- SourceAddressPrefix(es) / DestinationAddressPrefix(es)
- SourceApplicationSecurityGroups(s) / DestinationAddressSecurityGroup(s)

Would it make sense to accommodate for these use cases by logging a warning instead of returning an error? This would result on a post-create failure while trying to set up Docker instead of a on create failure.